### DISCUSSION TITLE
Caching per user

### DISCUSSION THREAD

**Discussion by tobyl** [vercel/swr#1713] [Q&A] [answered]
This is less of a question about SWR and more about caching generally in browsers and with API calls, so apologies in advance is this seems unrelated to SWR. I have a user auth setup where I need to fetch a user status and then begin a front end session once the status is received - along the lines of:

 ```
const { data: statusData, error: statusError } = useSWR('/users/status/', fetcher, {
  onErrorRetry: (error, key, config, revalidate, { retryCount }) => {
    if (error.response.status !== 202) return
    if (retryCount >= 15) return
    setTimeout(() => revalidate({ retryCount }), 1000)
  }
})

const { data: sessionData, error: sessionError } = useSWR(
  (statusData && statusData.refresh_status === 'complete') ? '/users/session/status/' : null, fetcher,
  { refreshInterval: 1000 * 10 },
)

// a bunch of other data calls dependent upon the result of these two calls
```

This is admittedly somewhat obscure, but the idea is that the 2nd call is only triggered if the first call resolves with a specific value. I tested pretty thoroughly and it works.

In production however I'm seeing some issues where a user logging into one account and then logging into a different account will see data from another user's session. I understand this is due to the browser (and possibly SWR?) caching the response from the urls since they are the same for both users.

In previous systems I would have "cache busted" the calls by adding something like `?timestamp=<current-date-time>` but I don't want to negate the benefits of using SWR.

What would be the recommended way to solve this issue? I can append a user-specific string to the URLs instead of using the current date and time, but should this be passed as part of the URL or can I include it as a value in the key array?

Apologies again for the obscure and most likely off-topic nature of this question, and thanks in advance!

**Comment by huozhi** [comment, marked as answer]
If you're be able to fetch user status I'd assume you almost have a user id like identifier? you can use the endpoint and id together as a key but only pass the endpoint to fetcher. I think in that way each cache state per user will use different cache

**Reply by tobyl** [reply]
Thank you for the response, this is helpful - this is mentioned in the docs, if not specifically as a solution to caching per user. My only hesitation is that I'd have to add this for every call and then test/troubleshoot each call. Is this something I could configure in the global config as some kind of key modifier? Or would this be a good use case for middleware? Any recommendations would be greatly appreciated.

**Reply by shuding** [reply]
You can use the cache provider feature to separate the cache by namespace: https://swr.vercel.app/docs/advanced/cache#create-cache-provider simply by using a Map (user ID → cache provider).

**Reply by tobyl** [reply]
That sounds like exactly what I'm looking for - thank you so much!
